feat: i18n operator names

Constrat opened the pull request:
ref: https://github.com/ZOOT-Plus/zoot-plus-frontend/commit/8196dc42a86ef997c960f6ae7d89a11bb1fb67bf#commitcomment-156785884

<details>
<summary>Resolved</summary>

As usual. This is still very wip and early implmentation just to see what to do on the backend first. Implemeting the global repo is much simpler and scalable enough for other languages if necessary, instead of completely changing everything to use the battle_data names.


### Current issues:
Unreleased operators on the global don't appear... Duh, the gamedata doesn't have them so the fallback is not really working currently lol.

### Missing:
- src/components/editor/**

</details>

Constrat commented:
![{F315681B-C07B-480C-949D-F2D1FB02BCC5}](https://github.com/user-attachments/assets/5f9ad144-5355-4254-b7a5-63efc4a75c47)
![{9D096496-0028-4BBB-B3E4-A230CD3BF4E8}](https://github.com/user-attachments/assets/01cccb85-6f99-4979-a054-158a3230e27d)

Still, looks crisp already XD

As usual long names are a pain in the ass.

guansss commented:
There are a lot of duplicate informations in `operators_en.json` that unnecessarily increase the bundle size.

Since only the operator names and profession names are dependent on languages (`alias` and `alt_name` are not needed for other languages currently), we can just put them in the original `operators.json` with something like `"name"` and `"name_en"`, just like how MAA did in `battle_data.json`.

> Unreleased operators on the global don't appear

Can be solved by searching over both `"name"` and `"name_en"`, I guess?

> As usual long names are a pain in the ass.

Well yeah, we need to redesign these UIs at some point.

guansss commented:
Could you check if CN jobs work in EN client and if EN jobs work in EN client? Through my tests EN jobs do not seem to work in CN client.

Constrat commented:
Ok! I will make the necessary changes. At first I made two jsons for simplicity and to just see how it would look. I definitely agree that we just need the localized name in the main JSON.

Constrat commented:
> Could you check if CN jobs work in EN client and if EN jobs work in EN client? Through my tests EN jobs do not seem to work in CN client.

I do not understand this. You mean the json itself? What does the translation have to do with anything? Does it "create" the json from the frontend displayed operator names? Meaning it will have english names in the json? I tried downloading a json and it seemed correct.

Constrat commented:
Also, we were pretty close in having to do nothing to the gamedata extractio since `alt_name` uses the "english" name already but it's not exactly the official one :( 
![{058C9C55-E893-4B23-A697-08AFEE69E9E6}](https://github.com/user-attachments/assets/52873fbe-e8ab-4d18-933a-6fc37495aa1d)
![{58E82C98-865B-4E7A-8267-450CDEFFF700}](https://github.com/user-attachments/assets/8f93eeac-67fb-4b9b-94d9-fcf5b4e4a70d)

Constrat commented:
@guansss I've been working on trying to change the logic and lighten everything up. Looking at copilot.schema.ts and how the names are called in all components. We are calling the name directly yes? I thought we were using IDs. This makes it so that we have to go from cn name -> id -> english name, instead of going directly to cn or en from the id. I don't know if I made myself clear.

Constrat commented:
Every time OperatorInfo is called, this will require adding name_en as empty I'm guessing?

guansss commented:
> Could you check if CN jobs work in EN client and if EN jobs work in EN client? Through my tests EN jobs do not seem to work in CN client.

Never mind this. I got some misinformation while testing.

The fact is, when MAA is executing the json, it always converts the recoginized operator names to Chinese before matching them against the names in json, meaning that the names in json should always be in Chinese regardless of the language of the game client to run in.

According to that, we don't need to change the current behavior of generating the json with Chinese names (via `operator.name`). And the i18n operator names should only be used for displaying on the website.

> Or change all components to use the a helper function that based on the chinese name will get the english one?

I think this is the way to go.

> Also, we were pretty close in having to do nothing to the gamedata extractio since alt_name uses the "english" name already but it's not exactly the official one :(

Yeah, the `appellation`s in the CN game data are basically the same with official English names with a few (likely 50?) exceptions. But `appellation`s also cover unreleased operators in the global server, so I suggest using them as a fallback source.

I imagine something like:

```js
// pseudo code
function getEnglishName(id) {
  const cnOp = findCnOp(id)
  const enOp = findEnOp(id)

  if (enOp) {
    return enOp.name
  }
  return cnOp.appellation
}
```

> Every time OperatorInfo is called, this will require adding name_en as empty I'm guessing?

I'm not sure what this means?

Constrat commented:
![{4DB09435-BC83-432B-84F3-DE45C44D690A}](https://github.com/user-attachments/assets/42aa4511-f5ef-47ec-8e89-2d852985b88a)
Here since we use OperatorInfo (OperatorInfo comes from OPERATORS) since we added `name_en` to the json, it requires the name_en variable.

Constrat commented:
![{123E5A7B-6404-4672-A248-C34077B92D44}](https://github.com/user-attachments/assets/bfa35e2b-4613-4967-9572-574025b0f1df)
Oh. This is a problem. I'm very not versed in CSS

Constrat commented:
- Added `getLocalizedOperatorName` function to retrieve operator names based on the selected language.
- Added `useLocalizedOperatorName` hook to utilize the localization in components.

guansss commented:
![image](https://github.com/user-attachments/assets/40a536bd-51e2-4774-9efd-32a641634699)

I've tweaked the CSS and it looks better now?

As for the `useLocalizedOperatorName` hook, its usage actually violates React's [Rules of Hooks](https://react.dev/reference/rules/rules-of-hooks). I would suggest simply using `getLocalizedOperatorName(name, lang)` in place of it.

PS: It looks like you don't have an ESLint extension installed in your IDE as there are several lint errors in the code. You can install one, or manually run `yarn lint` / `yarn lint:fix` before committing the code to see / fix the errors.

Constrat commented:
CSS looks very nice!

I'll work on the linting then

Constrat commented:
Oh, ESLint is pretty cool. In-line prettier is very useful.

guansss commented:
LGTM!

Constrat commented:
🫡
